stl/src/vector_algorithms.cpp

```diff
@@ -2037,161 +2038,375 @@ namespace {
         }
+
 #endif // !_M_ARM64EC
 
-        auto _Ptr_haystack           = static_cast<const _Ty*>(_First1);
-        const auto _Ptr_haystack_end = static_cast<const _Ty*>(_Last1);
-        const auto _Ptr_needle       = static_cast<const _Ty*>(_First2);
-        const auto _Ptr_needle_end   = static_cast<const _Ty*>(_Last2);
+        template <class _Traits>
+        const void* __stdcall __4_8_impl(const void* const _First1, const void* const _Last1, const void* const _First2,
+            const void* const _Last2) noexcept {
+            using _Ty = _Traits::_Ty;
+#ifndef _M_ARM64EC
+            if (_Use_avx2()) {
+                _Zeroupper_on_exit _Guard; // TRANSITION, DevCom-10331414
+
+                const size_t _Needle_length = _Byte_length(_First2, _Last2);
+                const int _Needle_length_el = static_cast<int>(_Needle_length / sizeof(_Ty));
 
-        for (; _Ptr_haystack != _Ptr_haystack_end; ++_Ptr_haystack) {
-            for (auto _Ptr = _Ptr_needle; _Ptr != _Ptr_needle_end; ++_Ptr) {
-                if (*_Ptr_haystack == *_Ptr) {
-                    return _Ptr_haystack;
+                // Special handling of small needle
+                // The generic approach could also handle it but with worse performance
+                if (_Needle_length_el == 0) {
+                    return _Last1;
+                } else if (_Needle_length_el == 1) {
+                    _STL_UNREACHABLE; // This is expected to be forwarded to 'find' on an upper level
+                } else if (_Needle_length_el == 2) {
+                    return __shuffle_impl<_Traits, 2>(_First1, _Last1, _First2, _Needle_length_el);
+                } else if (_Needle_length_el <= 4) {
+                    return __shuffle_impl<_Traits, 4>(_First1, _Last1, _First2, _Needle_length_el);
+                } else if (_Needle_length_el <= 8) {
+                    if constexpr (sizeof(_Ty) == 4) {
+                        return __shuffle_impl<_Traits, 8>(_First1, _Last1, _First2, _Needle_length_el);
+                    }
                 }
-            }
-        }
 
-        return _Ptr_haystack;
-    }
+                // Generic approach
+                const size_t _Needle_length_tail = _Needle_length & 0x1C;
+                const __m256i _Tail_mask         = _Avx2_tail_mask_32(_Needle_length_tail >> 2);
+
+                const void* _Stop2 = _First2;
+                _Advance_bytes(_Stop2, _Needle_length & ~size_t{0x1F});
 
+                for (auto _Ptr1 = static_cast<const _Ty*>(_First1); _Ptr1 != _Last1; ++_Ptr1) {
+                    const auto _Data1 = _Traits::_Set_avx(*_Ptr1);
+                    for (auto _Ptr2 = _First2; _Ptr2 != _Stop2; _Advance_bytes(_Ptr2, 32)) {
+                        const __m256i _Data2 = _mm256_loadu_si256(static_cast<const __m256i*>(_Ptr2));
+                        const __m256i _Eq    = _Traits::_Cmp_avx(_Data1, _Data2);
+                        if (!_mm256_testz_si256(_Eq, _Eq)) {

```

Comment:
`_mm256_testz_si256` returns `int`, so this is directly negating an integer, which we conventionally avoid. Occurs below.

Reply:
That function is documented to only return 0 and 1, so personally I'd say it should be treated as a boolean. It's probably typed as int because there is no bool type in the C89 language.

But it's your repo, your rules.

Reply:
After looking at this again, I agree with you - the direct testing is clearer to understand (avoids getting confused with layers of negation). I'll revert this commit. Thanks!